File: Day2/Python/Main.py
import os

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    inputFile = open(input, 'r')
    correctPwds1 = 0
    correctPwds2 = 0
    for line in inputFile.readlines():
        inputs = line.split(' ')
        range = inputs[0].split('-')
        min = int(range[0])
        max = int(range[1])
        occurrences = inputs[2].count(inputs[1][0])
        if occurrences >= min and occurrences <= max:
            correctPwds1 += 1

        if (inputs[2][min - 1] == inputs[1][0]) != (inputs[2][max - 1] == inputs[1][0]):
            correctPwds2 += 1

    print(f"Found {correctPwds1} correct step 1 passwords.")
    print(f"Found {correctPwds2} correct step 2 passwords.")

if __name__ == '__main__':
    main()

File: Day2/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		int correctPwds1 = 0;
		int correctPwds2 = 0;
		for (String line : Files.readAllLines(input.toPath())) {
			String[] inputs = line.split(" ");
			String[] range = inputs[0].split("-");
			int min = Integer.parseInt(range[0]);
			int max = Integer.parseInt(range[1]);
			char requested = inputs[1].charAt(0);
			int occurrences = (int) inputs[2].chars().filter(c -> c == requested).count();
			if (occurrences >= min && occurrences <= max) {
				correctPwds1++;
			}

			if ((inputs[2].charAt(min - 1) == requested) != (inputs[2].charAt(max - 1) == requested)) {
				correctPwds2++;
			}
		}

		System.out.println("Found " + correctPwds1 + " correct step 1 passwords.");
		System.out.println("Found " + correctPwds2 + " correct step 2 passwords.");
	}

}
